File: simple_factory/typescript/simple_factory.ts
interface Veiculo {
    buscarCliente(): void;
}

class CarroLuxo implements Veiculo {
    buscarCliente() {
        console.log('Carro de luxo buscando cliente');
    }
}

class CarroPopular implements Veiculo {
    buscarCliente() {
        console.log('Carro popular buscando cliente');
    }
}

class Moto implements Veiculo {
    buscarCliente() {
        console.log('Moto buscando cliente');
    }
}

class VeiculoFactory {
    private carro: Veiculo;

    constructor(tipo: string) {
        this.carro = this.getCarro(tipo);
    }

    private getCarro(tipo: string): Veiculo {
        switch (tipo) {
            case 'luxo':
                return new CarroLuxo();
            case 'popular':
                return new CarroPopular();
            case 'moto':
                return new Moto();
            default:
                throw new Error('Tipo de veículo desconhecido');
        }
    }

    buscarCliente() {
        this.carro.buscarCliente();
    }
}

function main() {
    // Coloque no parâmetro do VeiculoFactory o tipo de veículo que deseja
    const veiculo = new VeiculoFactory('luxo');
    veiculo.buscarCliente();
}

main();

File: simple_factory/python/simple_factory.py
from abc import ABC, abstractmethod


class Veiculo(ABC):
    @abstractmethod
    def buscar_cliente(self) -> None: pass

class CarroLuxo(Veiculo):
    def buscar_cliente(self) -> None:
        print('Carro de luxo está buscando o cliente...')

class CarroPopular(Veiculo):
    def buscar_cliente(self) -> None:
        print('Carro popular está buscando o cliente...')

class Moto(Veiculo):
    def buscar_cliente(self) -> None:
        print('Moto está buscando o cliente...')


class VeiculoFactory:
    def __init__(self, tipo):
        self.carro = self.get_carro(tipo)

    @staticmethod
    def get_carro(tipo: str) -> Veiculo:
        if tipo == 'luxo':
            return CarroLuxo()
        if tipo == 'popular':
            return CarroPopular()
        if tipo == 'moto':
            return Moto()
        assert 0, 'Veículo não existe'

    def buscar_cliente(self) -> None:
        self.carro.buscar_cliente()

    

if __name__ == '__main__':
    from random import choice
    veiculos_disponiveis = ['luxo', 'popular', 'moto']
    for _ in range(10):
        veiculo = VeiculoFactory(choice(veiculos_disponiveis))
        veiculo.buscar_cliente()




    